contracts/example/ZoraWrapper/ZoraWrapper.sol

```diff
@@ -0,0 +1,241 @@
+            "ZoraWrapper: Only owner or protocol can unwrap"
+        );
+
+        // If some token price is not know yet, update it now
+        if (pendingTokenId != 0) updatePendingTokenPrice(pendingTokenId);
+
+        uint256 priceToPay = price[_tokenId];
+
+        // Delete price and pendingTokenId to prevent reentrancy
+        delete price[_tokenId];
+        delete pendingTokenId;
+
+        // transfer voucher to voucher owner
+        IERC721(voucherAddress).safeTransferFrom(address(this), ownerOf(_tokenId), _tokenId);
+
+        // Transfer token to protocol
+        if (priceToPay > 0) {
+            // No need to handle native separately, since Zora Auction House always sends WETH
+            IERC20(cachedExchangeToken[_tokenId]).safeTransfer(protocolAddress, priceToPay);
+        }
+
+        delete cachedExchangeToken[_tokenId]; // gas refund
+
+        // Burn wrapped voucher
+        _burn(_tokenId);
+    }
+
+    /**
+     * @notice Handle transfers out of Zora Auction House.
+     *
+     * @param _from The address of the sender.
+     * @param _to The address of the recipient.
+     * @param _tokenId The token id.
+     */
+    function _beforeTokenTransfer(
+        address _from,
+        address _to,
+        uint256 _tokenId
+    ) internal virtual override(ERC721) {
+        if (_from == zoraAuctionHouseAddress && _to != address(this)) {
+            // Auction is over, and wrapped voucher is being transferred to voucher owner
+            // If recipient is address(this), it means the auction was canceled and price updating can be skipped
+
+            // If some token price is not know yet, update it now
+            if (pendingTokenId != 0) updatePendingTokenPrice(pendingTokenId);
+
+            // Store current balance and set the pending token id
+            price[pendingTokenId] = getCurrentBalance(_tokenId);
+            pendingTokenId = _tokenId;
+        }
+
+        super._beforeTokenTransfer(_from, _to, _tokenId);
+    }
+
+    function updatePendingTokenPrice(uint256 _tokenId) internal {
+        price[pendingTokenId] = getCurrentBalance(_tokenId) - price[pendingTokenId];
+    }
+
+    /**
+     * @notice Gets own the token balance for the exchange token, associated with the token ID.
+     *
+     * @dev If the exchange token is not known, it is fetched from the protocol and cached for future use.
+     *
+     * @param _tokenId The token id.
+     */
+    function getCurrentBalance(uint256 _tokenId) internal returns (uint256) {
+        address exchangeToken = cachedExchangeToken[_tokenId];
+
+        // If exchange token is not known, get it from the protocol.
+        if (exchangeToken == address(0)) {
+            uint256 offerId = _tokenId >> 128; // OfferId is the first 128 bits of the token ID.
+
+            // Get Boson offer. Don't explicitly check if the offer exists, since existance of the token implies it does.
```

Comment:
> Don't explicitly check if the offer exists, since existence of the token implies it does

if `_tokenId` is a token id previous to `v2.2.0` ,`offerId` here will be 0 here and fetching `getOffer(0)` would return `[exist: false, offer: [...offer, exchangeToken: addressZero]` so we would assume exchangeToken as being `wethAddress` when actually we don't know if this is the correct exchangeToken as we fetch offer 0 and not true offer id. Does that make sense? maybe I'm missing something,